Armadillo/src/main/java/com/scribd/armadillo/error/ArmadilloException.kt

```diff
@@ -2,84 +2,103 @@ package com.scribd.armadillo.error
 
 import com.scribd.armadillo.actions.Action
 
-sealed class ArmadilloException(exception: Exception) : Exception(exception) {
+sealed class ArmadilloException(cause: Throwable? = null,
+                                message: String = "Unhandled Armadillo Error")
+    : Exception(message, cause) {
     abstract val errorCode: Int
 }
 
 /**
  * Errors in Armadillo
  */
-data class EngineNotInitialized(val reason: String) : ArmadilloException(exception = Exception(reason)) {
+class EngineNotInitialized(message: String)
```

Comment:
why not make these data classes as well like
```
data class EngineNotInitialized(val reason: String)
    : ArmadilloException(message = "Armadillo Engine is not initialized: $reason") {
    override val errorCode = 100
}
```

Reply:
adding a data class wouldn't add anything, since `cause` and `message` are both already values in the superclass. 

adding some getters and setters wouldn't really change anything.

Reply:
I'm not really talking about getters and setters. The equality from `equals` would help us resolve the states of immutable objects if the updates are posted to a `conflated` objects like `StateFlows` which internally would use equals method and subside unnecessary emits thus helping us with view redraws or collections is what I think we would benefit from in the long run by using data classes

Reply:
These are `Exception`s. They are for error handling, and they need to conform to norms and expectations set by other `Exception` s.  Armadillo already has a data class,`ErrorAction` which may be what is being confused with here.  `Exceptions` aren't directly for `StateFlows`, these are meant to be raised and caught. And if they *aren't* then I would like to know so I can fix that too.

Furthermore `equals` wouldn't change anything either, because there are no `vals` exclusive in the data class separate from the superclass's equals method.

Reply:
ok thanks this is helpful